### DISCUSSION TITLE
Additional extensions made on the HAPI for the FHIR Core & Web usecase.

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#1989] [Announcements]
**Context**
---
- FHIR Core & Web are HAPI FHIR mobile and web clients respectively.  For these clients to work well we needed to add a few extensions to the existing HAPI APIs. 
- These extensions include
   - API authentication and Authorization using KeyCloak.
   - Generation of a Location Hierarchy. 
   - Generation of the Practitioner Details endpoint. 
 - This is a summary of the extensions we have added to the HAPI FHIR Server. 

**Extensions summary.** 
---

1. **Authentication using KeyCloak**. 
    - We use KeyCloak for Identity and Access Management.
    - We needed to have the HAPI FHIR APIs behind the KeyCloak server to make sure that only authorized individuals access the data on the HAPI instance. 
    - The extension code can be found [here](https://github.com/opensrp/hapi-fhir-keycloak) 
    - Here is a sample implementation on the JPA server 
       - https://github.com/opensrp/hapi-fhir-jpaserver-starter/blob/2fdee43c4afa5632ebdbf35a2f90b4d4b2e0cf72/pom.xml#L152-L156
       - https://github.com/opensrp/hapi-fhir-jpaserver-starter/blob/2fdee43c4afa5632ebdbf35a2f90b4d4b2e0cf72/src/main/resources/application.yaml#L201-L207
2. **Location Hierarchy endpoint.** 
    - Returns all the locations in a hierarchy format. 
    - It uses the `partOf` attribute in the Location resource to create the location hierarchy. 
    - The extension code can be found [here](https://github.com/opensrp/hapi-fhir-opensrp-extensions/tree/main/location)
    - Here is a sample implementation on the JPA server
        - https://github.com/opensrp/hapi-fhir-jpaserver-starter/blob/2fdee43c4afa5632ebdbf35a2f90b4d4b2e0cf72/pom.xml#L159-L163
    - Here is a sample payload https://github.com/opensrp/web/issues/841 
3. **Practitioner details endpoint** 
    - Gets all the Resources linked to a Practitioner.  
    - It is used to identify where the practitioner is assigned. 
    - https://github.com/opensrp/fhircore/issues/853 
    - The extension code can be found [here](https://github.com/opensrp/hapi-fhir-opensrp-extensions/tree/main/practitioner)
    - Here is a sample implementation on the JPA server
        - https://github.com/opensrp/hapi-fhir-jpaserver-starter/blob/2fdee43c4afa5632ebdbf35a2f90b4d4b2e0cf72/pom.xml#L165-L169

**Comment by ageryck** [comment]
Added a client-facing doc of the same [here](https://docs.google.com/document/d/1KUhkqZaBfJ4LT3MFx0TgNHpeKnH0sizNjwEy-eucQqg/edit#)
